ecs_tag and integration tests for new feature request

mpechner opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Wrote module ecs_tag and the integration tests

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->

- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ecs_tag implements the boto3 ECS tag_resource() and untag_resource()
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```
https://github.com/ansible/ansible/issues/51122

mpechner committed:
Fork take THREE

mpechner committed:
got screwed accidentally having a conflict file and not the real deal.

mpechner committed:
Reworked. last rebase attempt was a bad experience.

mpechner committed:
Added integrations test for ecs_tag

mpechner commented:
shipit

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+        return boto3_tag_list_to_ansible_dict(ecs.list_tags_for_resource(resourceArn=resource)['tags'])
+    except (BotoCoreError, ClientError) as e:
+        module.fail_json_aws(e, msg='Failed to fetch tags for resource {0}'.format(resource))
+    except InvalidParameterException as e:
```
If I remember correctly InvalidParameterException is a ClientError, so you wouldn't reach this.  You'd need to test for the InvalidParameterException before the other two...

mpechner replied:
I'll pull that out.  For some reason it was not printing the parameter.

tremble commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,331 @@
+- block:
+    - name: set up aws connection info
+      set_fact:
```
This is better done with "module defaults": https://docs.ansible.com/ansible/latest/dev_guide/platforms/aws_guidelines.html#aws-credentials-for-integration-tests

mpechner replied:
Will read it and update the tests.  I copied the  code from the ecs_cluster tests.

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,238 @@
+    ecs = module.client('ecs')
+
+    try:
+        if resource_type == 'cluster':
```
Something feels kinda wrong about this if / elif / elif / elif .

I believe it's slightly more efficient (and readable) to use Python's exception handling capabilities rather than all of your if len()/ if ...in...  statements  (it's slow to 'fail' but fast to succeed, which given we then error out is fine).

```
try:
    if resource_type == 'cluster':
        description = ecs.describe_clusters(clusters=[resource])
        resource_arn = description['clusters'][0]['clusterArn']
    elif resource_type == 'task':
        description = ...
        resource_arn = ...
...
except (IndexError, KeyError):
    module.fail_json(msg='Failed to find {0} {1}'.format(resource_type, resource))
except (BotoCoreError, ClientError) as e:
    module.fail_json_aws(e, msg='Failed to find {0} {1}'.format(resource_type, resource))
```

mpechner replied:
Yes, Makes sense, It also shortens main down quite a bit.  But I can still move that section to a function if you prefer.

tremble replied:
I'd say pull it out into its own function still, it helps make it clear what each section is doing.

tremble reviewed (changes requested):
Mostly stylistic, but I kinda dislike how long main() is.

Breaking out the try/except structure used to fetch the ARNs might help.

mpechner committed:
Except for the move the module_defaults I made the changers tremble suggested.
Added the extra test.  If resource does not exist.

jillr commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,235 @@
+- name: Remove all tags except for Name from a cluster
+  ecs_tag:
+    cluster_name: mycluster
+    resource_type: cluster:27
```
```suggestion
    resource_type: cluster
```

mpechner replied:
looks like a fat finder to me done

jillr commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,350 @@
+        state: present
+        <<: *aws_connection_info
+      register: ecs_taskdefinition_creation
+    - name: print task def info
```
Can you please remove debug tasks from the tests?

mpechner replied:
all gone

jillr commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,350 @@
+      assert:
+        that:
+          - taglist.failed == true
+          - taglist.changed == false
```
Can you please add a assertion for the contents of the failure message, to be sure we're failing in the expected way/place?

mpechner replied:
yup

jillr commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,350 @@
+        <<: *aws_connection_info
+      register: taglist
+
+    - name: cluster tags - No change after dding again
```
```suggestion
    - name: cluster tags - No change after adding again
```

mpechner replied:
thanks

jillr commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,350 @@
+        that:
+          - taglist.changed == true
+          - taglist.added_tags.Name == "task_definition-{{ resource_prefix }}"
+
```
Also an idempotency check here

mpechner replied:
Add twice.  On initial check both add_tags and tags.
On 2nd check tags and change == false

jillr commented on test/integration/targets/ecs_tag/tasks/full_test.yml:
```diff
@@ -0,0 +1,350 @@
+        that:
+          - taglist.changed == true
+          - taglist.added_tags.Name == "service-{{ resource_prefix }}"
+
```
Can you add an idempotency check here please?

mpechner replied:
Add twice.  On initial check both add_tags and tags.
On 2nd check tags and change == false

jillr commented on test/integration/targets/ecs_tag/tasks/main.yml:
```diff
@@ -0,0 +1,25 @@
+- set_fact:
```
This virtual env shouldn't be needed, the requirements should be satisfied by the Ansible test container.  full_test.yml could be renamed to main.yml.

mpechner replied:
removed old main and renamed full_test  I copied the ecs_cluster code.

jillr reviewed (commented):
A few nits, tests are currently passing for me locally - great progress @mpechner!

mpechner committed:
ecs_tag.py - removed fat finger
full_test.yml is now main.yml
removed debugs
Testing invalid cluster test for the expected message.
Add idempotency test is  for adding tests to service and task_definition.

All Jillr comments addressed.

mpechner committed:
added entry for ecs_tag

jillr commented on lib/ansible/config/module_defaults.yml:
```diff
@@ -389,6 +389,8 @@ groupings:
   - aws
   ecs_service_info:
   - aws
+  ecs_tag
```
```suggestion
  ecs_tag:
```

jillr commented on test/integration/targets/ecs_tag/tasks/main.yml:
```diff
@@ -0,0 +1,376 @@
+- block:
```
```suggestion
- module_defaults:
    group/aws:
      aws_access_key: '{{ aws_access_key | default(omit) }}'
      aws_secret_key: '{{ aws_secret_key | default(omit) }}'
      security_token: '{{ security_token | default(omit) }}'
      region: '{{ aws_region | default(omit) }}'

  block:
```

jillr commented on test/integration/targets/ecs_tag/tasks/main.yml:
```diff
@@ -0,0 +1,376 @@
+      ecs_cluster:
+        name: "{{ resource_prefix }}"
+        state: present
+        <<: *aws_connection_info
```
```suggestion
```
Once module_defaults is added, remove the yaml anchor from each task that uses it.

jillr reviewed (commented):
With the latest patchset and the noted changes things look good IMO.

mpechner committed:
Changed from old school facts and using  <<: *aws_connection_info to using module_defaults.
Thanks Jillr.

mpechner committed:
Missed a ':' and forgot to running sanity test that would have caught it.

mpechner committed:
Removed files I copied overe from the ecs_cluster test

tremble commented on lib/ansible/modules/cloud/amazon/ecs_tag.py:
```diff
@@ -0,0 +1,235 @@
+
+
+def get_arn(ecs, module, cluster_name, resource_type, resource):
+
```
**If** you make any additional changes I'd suggest removing this addition blank line (just for consistency)
```suggestion
```

mpechner replied:
If Jillr find anything else I'll make this change.

tremble reviewed (approved):
LGTM.  I've suggested a change but I'd really only bother applying it if you're making any other changes at the same time.

jillr reviewed (approved):
LGTM, thanks for this contribution @mpechner!

mpechner commented:
Excellent.

mpechner commented:
open a ticket against me.  I'll write the info module if you prefer.
